File: 2015/day2.rs
use std::fs;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();
  println!("Part One: {:?}", part_one(input.clone()));
  println!("Part Two: {:?}", part_two(input));
}

fn part_one(input: String) -> i32 {
  let mut sum = 0;
  for present in input.split("\n") {
    let mut dimensions: Vec<i32> = present
      .split("x")
      .map(|x| x.parse::<i32>().unwrap())
      .collect();
    sum += 2 * dimensions[0] * dimensions[1]
      + 2 * dimensions[1] * dimensions[2]
      + 2 * dimensions[0] * dimensions[2];
    dimensions.sort();
    sum += dimensions[0] * dimensions[1];
  }

  return sum;
}

fn part_two(input: String) -> i32 {
  let mut sum = 0;
  for present in input.split("\n") {
    let mut dimensions: Vec<i32> = present
      .split("x")
      .map(|x| x.parse::<i32>().unwrap())
      .collect();
    dimensions.sort();
    sum += 2 * dimensions[0] + 2 * dimensions[1];
    sum += dimensions[0] * dimensions[1] * dimensions[2];
  }
  return sum;
}

File: 2017/day2.ex
defmodule Main do
  use Application

  def start(_, _) do
    input =
      File.read!("input.txt")
      |> String.split("\n")
      |> Enum.map(fn line -> String.split(line, "\t") end)
      |> Enum.map(fn line -> Enum.map(line, fn x -> String.to_integer(x) end) end)

    IO.puts("Part one: #{Part_one.solve(input)}")
    IO.puts("Part two: #{Part_two.solve(input)}")

    Task.start(fn ->
      nil
    end)
  end
end

defmodule Part_one do
  def solve(input) do
    Enum.reduce(input, 0, fn line, acc -> acc + (Enum.max(line) - Enum.min(line)) end)
  end
end

defmodule Part_two do
  def solve(input) do
    Enum.reduce(input, 0, fn line, acc -> acc + divison_rec(Enum.sort(line, :desc)) end)
  end

  def divison_rec([head | tail]) do
    peter = Enum.find(tail, fn x -> rem(head, x) == 0 end)

    if(peter !== nil) do
      div(head, peter)
    else
      divison_rec(tail)
    end
  end
end
